cornflow-server/cornflow/shared/authentication/auth.py

```diff
@@ -195,65 +187,175 @@ def get_token_from_header(headers: Headers = None) -> str:
+                "Failed to fetch public keys from authentication provider",
+                log_txt=f"Error while fetching public keys from {jwks_url}: {str(e)}",
+                status_code=400
+            )
 
-        :param str token: the given token
-        :return: the user id code.
-        :rtype: int
+    def verify_token(self, token: str, provider_url: str, expected_audience: str) -> dict:
+        """
+        Verifies an OpenID Connect token
+        
+        :param str token: The token to verify
+        :param str provider_url: The base URL of the OIDC provider
+        :param str expected_audience: The expected audience claim
+        :return: The decoded token claims
+        :rtype: dict
         """
-        user_id = Auth.decode_token(token)["user_id"]
-        return user_id
+        # First validate token header
+        try:
+            unverified_header = jwt.get_unverified_header(token)
+        except jwt.InvalidTokenError:
+            raise InvalidCredentials(
+                "Invalid token format or signature",
+                log_txt="Error while verifying token. Invalid token format.",
+                status_code=400
+            )
 
-    """
-    START OF INTERNAL PROTECTED METHODS
-    """
+        # Check for kid in header
+        if "kid" not in unverified_header:
+            raise InvalidCredentials(
+                "Invalid token: Missing key identifier (kid) in token header",
+                log_txt="Error while verifying token. Token header is missing 'kid'.",
+                status_code=400
+            )
+            
+        kid = unverified_header["kid"]
+        
+        # Get public keys
+        public_keys = self.get_public_keys(provider_url)
+
+        
+        # Validate kid exists in public keys
+        if kid not in public_keys:
+            raise InvalidCredentials(
+                "Invalid token: Unknown key identifier (kid)",
+                log_txt=f"Error while verifying token. Key ID {kid} not found in public keys.",
+                status_code=400
+            )
+            
+        public_key = public_keys[kid]
+        
+        # Finally verify token
+        try:
+            payload = jwt.decode(
+                token,
+                public_key,
+                algorithms=["RS256"],
+                audience=expected_audience,
+                issuer=provider_url
+            )
+            return payload
+            
+        except jwt.ExpiredSignatureError:
+            raise InvalidCredentials(
+                "The token has expired, please login again",
+                log_txt="Error while verifying token. The token has expired.",
+                status_code=400
+            )
+        except jwt.InvalidTokenError as e:
+            raise InvalidCredentials(
+                "Invalid token format or signature",
+                log_txt=f"Error while verifying token signature: {str(e)}",
+                status_code=400
+            )
```

Comment:
And again, this is already present on the outer scope of where this is getting called. We can eb a little bit cleaner about this.